### ISSUE TITLE
Unable to add S3 object metadata on uploads

### ISSUE THREAD

**Issue by alan113696** [rclone/rclone#4303]
#### What is the problem you are having with rclone?
When uploading local files to S3, and using the `--header-upload` flag to specify S3 object metadata (e.g., "X-Amz-Meta-Test: FooBar"), I get the following error:

```
Failed to copy: s3 upload: 403 Forbidden: <?xml version="1.0" encoding="UTF-8"?><Error><Code>AccessDenied</Code><Message>There were headers present in the request which were not signed</Message><HeadersNotSigned>x-amz-meta-test</HeadersNotSigned><RequestId>removed</RequestId><HostId>removed</HostId></Error>
```

#### What is your rclone version (output from `rclone version`)
rclone v1.52.0
- os/arch: linux/amd64
- go version: go1.14.3

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
CentOS 7.8, 64-bit

####  Which cloud storage system are you using? (eg Google Drive)
Amazon S3

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
```
rclone copy test s3:mybucket/test -P --header-upload "X-Amz-Meta-Test: FooBar" --retries 1
```

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
```
2020/06/04 13:14:33 DEBUG : rclone: Version "v1.52.0" starting with parameters ["rclone" "copy" "test" "s3:mybucket/test" "-P" "--header-upload" "X-Amz-Meta-Test: FooBar" "--retries" "1" "-vv"]
2020/06/04 13:14:33 DEBUG : Using config file from "/home/myuser/.config/rclone/rclone.conf"
2020/06/04 13:14:33 DEBUG : fs cache: renaming cache item "test" to be canonical "/home/myuser/work/test"
2020-06-04 13:14:33 DEBUG : S3 bucket mybucket path test: Waiting for checks to finish
2020-06-04 13:14:33 DEBUG : S3 bucket mybucket path test: Waiting for transfers to finish
2020-06-04 13:14:34 ERROR : file1.txt: Failed to copy: s3 upload: 403 Forbidden: <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>There were headers present in the request which were not signed</Message><HeadersNotSigned>x-amz-meta-test</HeadersNotSigned><RequestId>removed</RequestId><HostId>removed</HostId></Error>
2020-06-04 13:14:34 ERROR : file2.txt: Failed to copy: s3 upload: 403 Forbidden: <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>There were headers present in the request which were not signed</Message><HeadersNotSigned>x-amz-meta-test</HeadersNotSigned><RequestId>removed</RequestId><HostId>removed</HostId></Error>
2020-06-04 13:14:34 INFO  : There was nothing to transfer
2020-06-04 13:14:34 ERROR : Attempt 1/1 failed with 2 errors and: s3 upload: 403 Forbidden: <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>There were headers present in the request which were not signed</Message><HeadersNotSigned>x-amz-meta-test</HeadersNotSigned><RequestId>removed</RequestId><HostId>removed</HostId></Error>
Transferred:            12 / 12 Bytes, 100%, 21 Bytes/s, ETA 0s
Errors:                 2 (retrying may help)
Elapsed time:         0.5s
2020/06/04 13:14:34 INFO  : 
Transferred:   	        12 / 12 Bytes, 100%, 21 Bytes/s, ETA 0s
Errors:                 2 (retrying may help)
Elapsed time:         0.5s

2020/06/04 13:14:34 DEBUG : 4 go routines active
2020/06/04 13:14:34 Failed to copy with 2 errors: last error was: s3 upload: 403 Forbidden: <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>There were headers present in the request which were not signed</Message><HeadersNotSigned>x-amz-meta-test</HeadersNotSigned><RequestId>removed</RequestId><HostId>removed</HostId></Error>
```

As an bonus, it would be useful to test the "x-amz-tagging" header to support S3 object tagging in addition to S3 object metadata. Then support both tagging and metadata on PutObject and Multipart Upload.

**Comment by alan113696** [comment]
[Reference to forum post](https://forum.rclone.org/t/unable-to-add-s3-object-metadata-on-uploads/16902)

**Comment by ncw** [comment]
Can you give this a go please? I think it should fix it for meta and tagging for single and multipart uploads.

https://beta.rclone.org/branch/v1.52.0-027-gd043ac64-fix-4303-s3-headers-beta/ (uploaded in 15-30 mins)

**Comment by alan113696** [comment]
The following works great for small files that do not use multipart. The two metadata and two tagging attributes are added to the objects.

`rclone-fix-s3-headers copy test s3:mybucket/test -P --header-upload "X-Amz-Meta-Test: FooBar" --retries 1 -vv --header-upload "x-amz-meta-yuk: junk" --header-upload "x-amz-tagging: foo=bar&hello=world"`

When a large file is added to test multipart, I get the following error.

```
panic: reflect: Field index out of bounds

goroutine 36 [running]:
reflect.(*structType).Field(0x19402a0, 0x1c, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	reflect/type.go:1185 +0x1e9
reflect.(*rtype).Field(0x19402a0, 0x1c, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	reflect/type.go:920 +0x7a
github.com/rclone/rclone/lib/structs.SetFrom(0x192e7c0, 0xc00033c000, 0x1933c80, 0xc000125000)
	github.com/rclone/rclone/lib/structs/structs.go:26 +0x2bc
github.com/rclone/rclone/backend/s3.(*Object).uploadMultipart(0xc00013c0e0, 0x1d4d460, 0xc0001dad80, 0xc000125000, 0x4475a28, 0x1d0fe40, 0xc000129040, 0x0, 0x0)
	github.com/rclone/rclone/backend/s3/s3.go:2209 +0x197
github.com/rclone/rclone/backend/s3.(*Object).Update(0xc00013c0e0, 0x1d4d460, 0xc0001dad80, 0x1d0fe40, 0xc000129040, 0x7f7ae444aa60, 0xc0000c3440, 0xc00007e0c0, 0x4, 0x4, ...)
	github.com/rclone/rclone/backend/s3/s3.go:2457 +0x68f
github.com/rclone/rclone/backend/s3.(*Fs).Put(0xc000626380, 0x1d4d460, 0xc0001dad80, 0x1d0fe40, 0xc000129040, 0x7f7ae444aa60, 0xc0000c3440, 0xc00007e0c0, 0x4, 0x4, ...)
	github.com/rclone/rclone/backend/s3/s3.go:1687 +0x11e
github.com/rclone/rclone/fs/operations.Copy(0x1d4d460, 0xc0001dad80, 0x1d63300, 0xc000626380, 0x0, 0x0, 0xc00029c497, 0x23, 0x1d62980, 0xc0000c3440, ...)
	github.com/rclone/rclone/fs/operations/operations.go:440 +0x179a
github.com/rclone/rclone/fs/sync.(*syncCopyMove).pairCopyOrMove(0xc0001a0380, 0x1d4d460, 0xc0001dad80, 0xc00021c0f0, 0x1d63300, 0xc000626380, 0x0, 0xc0001a0458)
	github.com/rclone/rclone/fs/sync/sync.go:367 +0x223
created by github.com/rclone/rclone/fs/sync.(*syncCopyMove).startTransfers
	github.com/rclone/rclone/fs/sync/sync.go:394 +0xd2
```

**Comment by ncw** [comment]
Ooops, not sure how that got through QA :blush: 

Hopefull this will work a bit better!

https://beta.rclone.org/branch/v1.52.0-032-gce9cfe10-fix-4303-s3-headers-beta/ (uploaded in 15-30 mins)

**Comment by alan113696** [comment]
Works great for multipart/non-multipart, object metadata, and object tagging!! Thanks for the fix!

**Comment by ncw** [comment]
> Works great for multipart/non-multipart, object metadata, and object tagging!! Thanks for the fix!

Thanks for testing :-)

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.52.1
